ykrt/src/trace/hwt/mapper.rs

```diff
@@ -1,37 +1,119 @@
 //! The mapper translates a hwtracer trace into an IR trace.
 
 use crate::trace::{AOTTraceIterator, AOTTraceIteratorError, InvalidTraceError, TraceAction};
-use hwtracer::llvm_blockmap::LLVM_BLOCK_MAP;
-use hwtracer::Trace;
-use std::error::Error;
+use hwtracer::{llvm_blockmap::LLVM_BLOCK_MAP, Block, HWTracerError, Trace};
 use ykaddr::{
     addr::{vaddr_to_obj_and_off, vaddr_to_sym_and_obj},
     obj::SELF_BIN_PATH,
 };
 
+/// Map the *machine* blocks of the specified trace into LLVM IR blocks.
+///
+/// Each entry in the returned trace is either a "mapped block" identifying a successfully
+/// mapped LLVM IR block, or an unsuccessfully mapped "unmappable block" (an unknown region of
+/// code spanning at least one machine block).
 pub(crate) struct HWTTraceIterator {
-    trace: std::vec::IntoIter<TraceAction>,
+    hwt_iter: Box<dyn Iterator<Item = Result<Block, HWTracerError>> + Send>,
+    /// The next [TraceAction]`s we will produce when `next` is called. We need this intermediary
+    /// to allow us to deduplicate mapped/unmapped blocks. This will be empty on the first
```

Comment:
> deduplicate mapped/unmapped

Do we de-duplicate mapped blocks too? I did not know that (or I did once, but have forgotten that detail).

Reply:
You implemented it :p https://github.com/ykjit/yk/blob/master/ykrt/src/trace/hwt/mapper.rs#L190

Reply:
OK, we really need to put that comment back in :)

Reply:
See https://github.com/ykjit/yk/pull/1008#discussion_r1512524230 and b95786d4feaea78efb099575ff02bd0e055ff4e6.

Reply:
Thanks